### ISSUE TITLE
Fully Implement Skeletal Animation

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#15]
**Another one from my scratch notes**

Task: Implement Bone Animation
------------------------------
Goal: Support animated models. At the moment, the bone animation kernel runs, and technically it does animate the hulls based on bones. However, the bone transforms in the buffer never change, so the animation simply keeps the model in the default bind-pose. The model will need some work to add at least one animation, then the animation data needs to be read and stored in a reference buffer. Kernels may need to be updated to read from the reference buffer in order to perform the necessary linear interpolation between keyframes. Also, some mechanism to select an animation will need to be done. For the purposes of quick iteration, mapping the player model animation cycle to a key will suffice. The current frames can then be calculated from the correct keyframes for the current animation, and written into the bone buffer, allowing the animation kernel to finally cause the hulls to move.

High-level:

1. create animation in model
    - use blender to add at least one animation
2. update model loader to support animation data
    - will need some new methods to load the data
3. update kbm input system to switch animation
    - bind a key to cycle animations
3. update or add reference buffers to store animations 
    - will need new variables in memory buffers to hold reference animation data
4. update or add CL kernels to perform bone matrix keyframe lerp
    - take the current frame and calculate the appropriate bone positions 
5. ensure current kernel functions properly, resulting in animation
    - animation should play on a loop

**Comment by controllerface** [comment]
The work on the humanoid renderer went a long way toward making this one possible, with all the required bone buffers created and being used in the object creation code. 

My current plan is to try and write a bone animation kernel that is called once per tick, at the very beginning of the loop, and then the existing animation kernel will re-animate when it is called. This will make it so the parent chains for the bones will only need to be walked once per tick but the point animation kernel can be called multiple times per tick. This is necessary to ensure physics sub-steps work properly.

If I can make this work easily enough, I might look at possibly splitting the animation task off into a separate thread, and let it manipulate the bone buffer directly. The open CL API is documented as supporting thread safety for pretty much all actions accept for setting kernel arguments, so i am hopeful that this could work at the very least for an isolated case like this.

**Comment by controllerface** [comment]
Did a little bit of work on this today, it turns out that the bone weight values need to be normalized so they sum up to 1. With that done, I was able to test and confirm bone weights are working correctly 👍 

Now starts the somewhat arduous process of teasing out what other memory structures are going to be needed, and also just getting some small animation working to test them. To start off, i am going to do a very basic animation, and I want to make sure only some bones are moved, since that is a case that is going to be very common, and should result in un-animated bones staying as they are in the bind pose.

I am unsure exactly how far I intended to go in this story, I want to make sure that animation works in terms of the kernels and app code works to ensure that an animation actually can play and loop. I also want to make sure there's some way to determine if an animation should loop or not, but I may make a second story out of that depending on the complexity.

**Comment by controllerface** [comment]
Did some work in blender, I now have a model with one very simple animation, just the humanoid model waving its arms up and down. I found that I needed to normalize the bone weights as they can sum up to more than 1. I still need a little more practice with weight painting to figure out exactly how this works, but it was easy enough to fix during model loading. 

I think I will need to add a new reference table for animations, and probably some related buffers to hold offsets of variable sized tables, etc. Armatures that are created will need to have a link to this reference data, and some way to store what animation is currently set. The idea is that a new kernel will be made that runs per-armature, which selects the proper animation, figures out what the current frame is using interpolation, and then updates the bone instances of the armature's constituent meshes. 

Looking ahead a bit, I am starting to wonder if I should figure out a mechanism for expanding buffers at runtime instead of the current design where I just grab buffers up front. I am still nowhere near taking up even 1 GB at the moment, so it's not pressing, but I can see how the animations could take up a large chunk of memory, and all things considered, having to pick a maximum aligned with other objects is probably not a good idea. This is out of scope for this ticket, but it is worth starting to think about.

**Comment by controllerface** [comment]
Little more tinkering last night and I now think I will need to add an armature aligned bone instance buffer to capture the current animation state for each bone instance, which can then be referenced. Initially I was planning on calculating the full parent hierarchy in the hull animation kernel, but after mulling this over a bit, it is both more complicated and likely less performant.

Instead, I'm going to trade some memory for a cleaner and easier to understand implementation. If I am careful in how I lay things out, I should be able to ensure parent objects are always _before_ child objects in the buffer, so I can do a single loop per armature to prepare the current bone animation state "recursively" by just going sequentially. After that kernel runs, the per hull kernel can then run, and reference the bone state buffer when animating. 

This will probably require some changes to the bone tables, but I think I can do it without expanding them, just changing what they are used for.

**Comment by controllerface** [comment]
OK, just going to jot these details here so I don't forget:

- add armature centric bind pose instance buffer
- extend armature hull table to int 4 to hold start/end for the new buffer
- update delete kernels to handle the new objects
   - new shift buffer will be needed
   - CPU code will also need to ensure correct buffer sizes
   - expand the current `int` edge buffer to `int2` 
- try and make the structure optional for simple meshes that won't have animations

**Comment by controllerface** [comment]
Took some doing, but finally have the buffers working, including the delete kernels. I now need to work on the kernel code that will update each of the armature bones for the current frame. Although, to get this working I can write the kernel assuming there is no animation for the current frame, and use the default bind pose. Once that functions properly, I can move on to actually making the animations update.

This will require some ordered traversals of objects to build up the parent chain. I have laid out the bindings in memory such that, as long as the higher level objects are first in order, the parent reference can be used directly to get the current combined offset at that level. This should allow following kernels to have the parent chain effect pre-computed, avoiding extra loops.

**Comment by controllerface** [comment]
OK, the new kernels are working as expected. The bones are being updated once per frame, and the output looks identical to what was seen before where the bones didn't update, which was the expected outcome. The difference is that the positions of the bones _are_ recalculated every frame, where before they simply never changed.

Now, the task at hand is actually storing the animation data in memory, and accessing it in the new armature animation kernel. Armatures will also need to be extended with extra info to store their current animation, as well as the last time step that was seen. This should be enough to get a simple looping animation to play.

**Comment by controllerface** [comment]
after some testing, I found that the kernels were not actually working as I had thought, but I have now gotten them _actually_ working. In doing so, i discovered one more item I need to store alongside armatures, the initial transform matrix. I mistakenly thought that using an identity matrix would work, but it turns out the initial transform applied to the first bone is actually not an empty matrix, but one that is derived from armature object itself. This matrix needs to be stored as a reference and made available to armatures. This should be easy enough to set up, with just some work needed to detect which transform is associated with the armature in the model loading code. Once I have this done then I can finally start working out how to store the animation data.

**Comment by controllerface** [comment]
I added the buffer for the model transform, and it works correctly. I started in on examining the animation data, and I am identifying more data that needs to be associated with armatures. In addition to the model transform, there will need to be some data that stores the current animation index, as well as the time of the last frame. These will be individual buffers as one will be an `int` and one will be a `float`. 

There will also need to be reference buffers for the animations themselves, and in addition, the animations may need to be linked to the model with a table. The keyframes will need to be aligned with the bone bind pose references and there will need to be some function that can select and interpolate the current keyframe for a bone.

**Comment by controllerface** [comment]
Took a little break but have started back in on figuring out buffer layout. This has been a little tricker than expected, but I'm getting there now.  They key thing to get around is that the objects that need to be stored have to be laid out in a way that works best for a GPU kernel, not CPU code. It has taken a bit of thought to figure out the best way to align everything, and it is not the same structure that works for CPU based animation code and therefore all the existing data on how to do it needs to be translated to "GPU land".

Rather than try to have everything entirely mapped out, I am starting with the basic key frame data, which itself has no dependency on any other piece of data, a parent object will "own" the key frame, but it is self-contained. Each of the three key types also will need a matching buffer, so that is 3 new buffers; 1 each for `position`, `rotation`, and `scaling` keys. Keys need their values stored as well as a timestamp marking what time the key is associated with. 

In my test model, there are the same number of keys for all 3 because of how I generated the animation, but this will not be assumed to be true. I did also notice that bones for which there were no animations had 2 keyframes, presumably a "forged" start and end frame, which have the exact same data. It seems that this may be added either by the model export function in blender or the Assimp library. Whichever the case, this neatly ensures that the issue of always needing at least 2 frames to interpolate between is not a concern.

Key frames that are stored are then going to be aligned with the named bone's base bind pose reference. At first, I was going to align all the animation data to the model in memory, as the table of animation indices, which is how most examples handle this. In the end though, this doesn't make a lot of sense for a GPU kernel, which operates on the bones directly. So instead, each bind pose reference will have aligned channel tables, which will hold indices into the key buffers.

With each key table having offsets that are computable based on the index values, it should be possible to calculate the two frames on either side of the current time step. Unfortunately, this means there will need to be loops for each, but I may manually unroll them to try and reduce their impact as much as possible. Once this is done, the interpolation needed should be easy enough to do using the existing example math functions I have.

**Comment by controllerface** [comment]
I have finally implemented most of the buffers using the approach of just starting with the most basic and working my way up. In the end, I was able to consolidate the keyframes to a single buffer pair, one for the keyframe data and one for the timestamp. The channels all use the same buffer, just pulling different ranges for the different types (position, rotation, and scaling).

I do still need to add armature-aligned buffers that hold the current animation and last rendered animation frame time, which means some more updates to the delete kernels, which is annoying but necessary. In addition, I need to pass the `dt` variable into the armature animation kernel so I can increment that last rendered time value on each time tick. 

After all that is done, I can then get to work on implementing the frame interpolation functions so I can finally start doing actual animations.

So, steady progress, but still a ways to go.

**Comment by controllerface** [comment]
It finally works!

This was so far the most complex feature to implement, but it was well worth it 😄 

I had originally thought about doing some more refinement to how animations are handled under this ticket, but I'm going to put it to bed and make new stories for follow up work, it's gone on long enough. I will probably do a little bit of cleanup before I finally merge it in, but I am pretty happy with what I have.

I will need to eventually learn how to make better animations too 😆
